src/state/opamFormatUpgrade.ml

```diff
@@ -1189,6 +1228,33 @@ let as_necessary requested_lock global_lock root config =
         config)
       config hard_upg
   in
+  let config =
+    let config, missing_switches = remove_missing_switches root config in
+    let local, global =
+      List.partition OpamSwitch.is_external missing_switches
+    in
+    let tostr sws =
+      OpamStd.Format.pretty_list
+        (List.map (OpamSwitch.to_string
+                   @> OpamConsole.colorise `bold
+                   @> Printf.sprintf "'%s'")
+           sws)
+    in
+    let sentences = function
+      | [_] -> "", "it no longer exists"
+      | _ -> "es", "they no longer exist"
+    in
+    (if not on_the_fly && global <> [] then
+       let switches, existence = sentences global in
+       OpamConsole.warning "Removing global switch%s %s as %s"
+         switches (tostr global) existence);
+    (if not on_the_fly && local <> [] then
+       let switches, existence = sentences local in
+       OpamConsole.note "Removing local switch%s %s as %s"
+         switches (tostr local) existence);
+    config
```

Comment:
No need for a message, it's silently handled in the normal (no update) case (it's an expected workflow that you just `rm -rf` source dirs).